kore/src/main/scala/org/kframework/definition/outer-ext.scala

```diff
@@ -57,90 +24,50 @@ case class Bubble(sentenceType: String, contents: String, att: Att = Att.empty)
-  def toModule(allModules: Set[FlatModule], koreModules: scala.collection.mutable.Map[String, Module], visitedModules: Seq[FlatModule]): Module = {
-    var items = this.localSentences
-    var importedModuleNames = this.imports
-    var importedSyntax = importedModuleNames.map(m => Import.asSyntax(m))
-
-    if (visitedModules.contains(this)) {
-      var msg = "Found circularity in module imports: "
-      visitedModules.foreach(m => msg += m.name + " < ")
-      msg += visitedModules.head.name
-      throw KEMException.compilerError(msg)
-    }
-
-    if (koreModules.contains(this.name))
-      return koreModules(this.name)
+object FlatModule {
+  def apply(name: String, unresolvedLocalSentences: Set[Sentence]): FlatModule = {
+    new FlatModule(name, Set(), unresolvedLocalSentences, Att.empty)
+  }
 
-    def resolveImport(_import: Import): Module = {
-      var baseName = Import.noSyntax(_import.name)
-      var modOption = allModules.find(m => m.name.equals(baseName))
-      if (modOption.nonEmpty) {
-        var mod = modOption.get
-        var result = koreModules.get(mod.name)
-        if (result.isEmpty) {
-          result = Some(mod.toModule(allModules, koreModules, this +: visitedModules))
-        }
-        if (Import.isSyntax(_import.name)) {
-          result = Some(koreModules.get(_import.name).get)
-        }
-          result.get
-      } else if (koreModules.contains(_import.name))
-          koreModules.get(_import.name).get
-        else
-          throw KEMException.compilerError("Could not find module: " + _import.name, _import)
+  /**
+   * Gets a list of {@link FlatModule} and returns a set of {@link Module}.
+   * @param allModules List of FlatModules to be transformed. The order matters when reporting circular imports.
+   * @param previousModules A set of Modules already built. New modules will be added to this set.
+   * @return The set of Modules, directly connected and maximally shared.
+   */
+  def toModules(allModules:Seq[FlatModule], previousModules:Set[Module]):Set[Module] = {
+    val memoization:mutable.HashMap[String, Module] = collection.mutable.HashMap[String, Module]()
+    previousModules.map(m => memoization.put(m.name, m))
+    def toModuleRec(m:FlatModule, visitedModules: Seq[FlatModule]):Module = {
+      if (visitedModules.contains(m)) {
+        var msg = "Found circularity in module imports: "
+        visitedModules.reverse.foreach(m => msg += m.name + " < ")
+        msg += visitedModules.last.name
+        throw KEMException.compilerError(msg)
+      }
+      memoization.getOrElseUpdate(m.name, {
+        val newM = new Module(
+          m.name,
+          m.imports.map(i => memoization.getOrElse(i.name,
+            toModuleRec(allModules.find(f => f.name.equals(i.name))
+              .getOrElse(throw KEMException.compilerError("Could not find module: " + i.name, i)), visitedModules :+ m))),
+          m.localSentences,
+          m.att
+        )
+        newM.checkSorts()
+        newM
```

Comment:
This reads a little gross to me, hard to understand the control flow because of the nested getOrElse* and the internal `map`.

Does it read better if you first compute the imports in a loop then construct the new module? Something like (psuedocode):

```suggestion
        val imports = Seq()
        for (String iName: m.imports) {
            if (! memoization.contains(iName) && ! allModules.contains(iName)) {
              throw KEMException.compilerError("Could not find module: " + i.name, i)
            } else {
              imports.add(memoization.getOrElseUpdate(iName, toModuleRec(allModules.find(iName), visitedModule :+ m))
            }
        }
        val newM = new Module(m.name, imports, m.localSentences, m.att)
        newM.checkSorts()
        newM
```

I dunno, it's just hard for me to unwrap the code as written.